File: src/main/java/com/oycm/algorithm/a/advance/Solution_7.scala
package com.oycm.algorithm.a.advance

object Solution_7 {

  /**
   * 1652.拆炸弹
   * https://leetcode.cn/problems/defuse-the-bomb/description/
   * 将 第 i 个数字按一下规则替换，得到一个新的结果：
   *  - k > 0，第 i 个数字值为当前之后 k 个数之和
   *  - k < 0，第 i 个数字值为当前之前 k 个数之和
   *  - k = 0，第 i 个数用 0 替换
   *
   * @param code 长度为 n 的环形数组
   * @param k
   * @return
   */
  def decrypt(code: Array[Int], k: Int): Array[Int] = {
    val n = code.length
    val ans = Array.fill(n)(0)
    var temp = 0

    if (k > 0) {
      for (i <- 1 until n + k) {
        // 入
        temp += code(i % n)
        // 窗口形成
        if (i >= k) {
          // 更多答案
          ans(i - k) = temp
          // 出 滑动
          // 怎么优化这里的 i-k+1 取模运算，只有一次会用到
          temp -= code((i - k + 1) % n)
        }
      }
    }

    if (k < 0) {
      for (i <- 0 until n - k - 1) {
        temp += code(i % n)
        if (i + k + 1 >= 0) {
          ans((i + 1) % n) = temp
          // 怎么优化这里的 i-k+1 取模运算，只有一次会用到
          temp -= code((i + k + 1) % n)
        }
      }
    }
    ans
  }

  /**
   * 计算优化：去掉取模，将原数组拼接
   * ans(0) = [l, r] sum; ans(1) = [l+1, r+1] sum
   *
   * @param code
   * @param k
   * @return
   */
  def decrypt_1(code: Array[Int], k: Int): Array[Int] = {
    val n = code.length
    val ans = Array.fill(n)(0)
    if (k == 0) {
      return ans
    }
    val newCode = Array.fill(2 * n)(0)
    Array.copy(code, 0, newCode, 0, n)
    Array.copy(code, 0, newCode, n, n)

    var l = if (k > 0) {
      1
    } else {
      n + k
    }
    var r = if (k > 0) {
      k
    } else {
      n - 1
    }
    var temp = 0
    for (i <- l to r) {
      temp += newCode(i)
    }
    for (i <- 0 until n) {
      ans(i) = temp
      temp -= newCode(l)
      temp += newCode(r + 1)
      l += 1
      r += 1
    }
    ans
  }

  def main(args: Array[String]): Unit = {
    println(decrypt(Array(5, 7, 1, 4), 3).toSeq == Array(12, 10, 16, 13).toSeq)
    println(decrypt(Array(1, 2, 3, 4), 0).toSeq == Array(0, 0, 0, 0).toSeq)
    println(decrypt(Array(2, 4, 9, 3), -2).toSeq == Array(12, 5, 6, 13).toSeq)

    println(decrypt_1(Array(5, 7, 1, 4), 3).toSeq == Array(12, 10, 16, 13).toSeq)
    println(decrypt_1(Array(1, 2, 3, 4), 0).toSeq == Array(0, 0, 0, 0).toSeq)
    println(decrypt_1(Array(2, 4, 9, 3), -2).toSeq == Array(12, 5, 6, 13).toSeq)
  }
}

File: src/main/java/com/oycm/algorithm/e/ea/advance/Solution_7.java
package com.oycm.algorithm.e.ea.advance;

public class Solution_7 {

    /**
     * 1031. <a href="https://leetcode.cn/problems/maximum-sum-of-two-non-overlapping-subarrays/description/">两个无重叠子数组的最大和</a> 1680
     * 找出长为 firstLen, secondLen 不重叠子数组的最大和
     *
     * @param nums
     * @param firstLen
     * @param secondLen
     * @return
     */
    public static int maxSumTwoNoOverlap(int[] nums, int firstLen, int secondLen) {
        /*
        暴力计算 [0, n - firstLen] 其对应可取 不重叠的 secondLen 的最大值，不断更新
         */
        int ans = 0;
        int n = nums.length;
        int first = 0;
        for (int i = 0; i < n; i++) {
            first += nums[i];
            int left = i - firstLen + 1;
            if (left >= 0) {
                // 计算 [0, left) 和 (i, n) 长为 secondLen 的最大值
                int sum = 0;
                int maxSum = 0;
                if (left >= secondLen) {
                    for (int j = 0; j < left; j++) {
                        sum += nums[j];
                        if (j - secondLen + 1 >= 0) {
                            maxSum = Math.max(sum, maxSum);
                            sum -= nums[j - secondLen + 1];
                        }
                    }
                }
                if (n - i - 1 >= secondLen) {
                    sum = 0;
                    for (int j = i + 1; j < n; j++) {
                        sum += nums[j];
                        // j - (i+1) + 1 >= secondLen
                        if (j - i >= secondLen) {
                            maxSum = Math.max(sum, maxSum);
                            sum -= nums[j - secondLen + 1];
                        }
                    }
                }
                ans = Math.max(ans, first + maxSum);
                first -= nums[left];
            }
        }


        return ans;
    }


    public static int answer(int[] nums, int firstLen, int secondLen) {
        /*
        题解思路: 前缀和
        firstLen 表示数组 a, 假设在左边, secondLen 表示数组在 右边
        a => [0, firstLen-1], b => [firstLen, firstLen + secondLen - 1], 每次移动，同时更新 a, b 数组的最大值
        i 在前缀和数组移动的范围 [firstLen, n - secondLen], 这里的 n 是 nums.length
        sa = s[i] - s[firstLen]
        sb = s[i+secondLen] - s[i]
        可以定义一个方法计算 左 a 右 b, 也就能计算 左 b 右 a 了
         */
        int n = nums.length;
        int[] s = new int[n + 1];
        for (int i = 0; i < n; i++) {
            s[i + 1] = s[i] + nums[i];
        }

        return Math.max(sum(s, firstLen, secondLen), sum(s, secondLen, firstLen));
    }

    public static int sum(int[] s, int firstLen, int secondLen) {
//        int a = 0;
//        int b = 0;
//        for (int i = firstLen; i < s.length - secondLen; i++) {
//            a = Math.max(a, s[i] - s[i - firstLen]);
//            b = Math.max(b, s[i + secondLen] - s[i]);
//        }
//        return a + b;
        // 注意 不能按上面的方式计算答案，上面的最大值数组会有交集 必须每次计算过程中更新答案
        int ans = 0;
        int maxA = 0;
        for (int i = firstLen; i < s.length - secondLen; i++) {
            maxA = Math.max(maxA, s[i] - s[i - firstLen]);
            ans = Math.max(ans, maxA + s[i + secondLen] - s[i]);
        }

        return ans;
    }


    public int optimize(int[] nums, int firstLen, int secondLen) {
        /*
        一次循环解决
         */
        int ans = 0;
        int n = nums.length;
        int[] s = new int[n + 1];
        for (int i = 0; i < n; i++) {
            s[i + 1] = s[i] + nums[i];
        }
        int maxA = 0;
        int maxB = 0;
        for (int i = firstLen + secondLen; i < n + 1; i++) {
            maxA = Math.max(maxA, s[i - secondLen] - s[i - secondLen - firstLen]);
            maxB = Math.max(maxB, s[i - firstLen] - s[i - secondLen - firstLen]);
            ans = Math.max(ans, Math.max(maxA + s[i] - s[i - secondLen], maxB + s[i] - s[i - firstLen]));
        }

        return ans;
    }

    public static void main(String[] args) {
        maxSumTwoNoOverlap(new int[]{8, 20, 6, 2, 20, 17, 6, 3, 20, 8, 12}, 5, 4);
        answer(new int[]{2, 1, 5, 6, 0, 9, 5, 0, 3, 8}, 4, 3);
    }
}
